Feat: Add experimental KV store

arv opened the pull request:
This implements the MemStore and IDBStore in JS.

The Rust side then wraps the store in a "proxy" that just calls the JS implementation.

Fixes #444

arv committed:
Playing around with implemening Store in JS

This only does MemStore at this point

arv commented:
Performance seems to be about the same:

**js-store**

```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 11.61 MB/s ±177.4% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 11.60 MB/s ±4.2% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 6.55 MB/s ±2.3% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 4.61 MB/s ±1.6% (5 runs sampled)
read tx 1024x1000 x 29.06 MB/s ±8.0% (15 runs sampled)
read tx 1024x5000 x 29.08 MB/s ±7.1% (5 runs sampled)
scan 1024x1000 x 57.78 MB/s ±17.2% (30 runs sampled)
scan 1024x5000 x 54.62 MB/s ±4.8% (6 runs sampled)
write single byte x 0.01 MB/s ±5400.0% (5697 runs sampled)
roundtrip write/subscribe/get x 2394.00 ops/sec ±1250.0% (1197 runs sampled)
create index 1024x1000 x 11.52 ops/sec ±1.3% (6 runs sampled)
create index 1024x5000 x 2.23 ops/sec ±0.8% (5 runs sampled)
subscription 10 x 139.97 ops/sec ±16.7% (70 runs sampled)
subscription 100 x 158.67 ops/sec ±40.3% (80 runs sampled)
subscription 1000 x 134.84 ops/sec ±76.8% (68 runs sampled)
Done!
```

**main**
```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 10.94 MB/s ±108.5% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 10.09 MB/s ±7.4% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 6.04 MB/s ±3.6% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 4.21 MB/s ±3.7% (5 runs sampled)
read tx 1024x1000 x 28.72 MB/s ±12.6% (15 runs sampled)
read tx 1024x5000 x 28.81 MB/s ±10.7% (5 runs sampled)
scan 1024x1000 x 55.49 MB/s ±15.9% (29 runs sampled)
scan 1024x5000 x 53.42 MB/s ±7.2% (6 runs sampled)
write single byte x Infinity MB/s ±Infinity% (9096 runs sampled)
roundtrip write/subscribe/get x 2451.51 ops/sec ±550.0% (1226 runs sampled)
create index 1024x1000 x 9.30 ops/sec ±1.9% (5 runs sampled)
create index 1024x5000 x 1.79 ops/sec ±0.7% (5 runs sampled)
subscription 10 x 135.14 ops/sec ±9.5% (68 runs sampled)
subscription 100 x 152.93 ops/sec ±18.5% (77 runs sampled)
subscription 1000 x 139.97 ops/sec ±16.9% (70 runs sampled)
Done!
```

arv commented:
Rust side: https://github.com/rocicorp/repc/pull/366

arv committed:
Add IDBStore in JS

arv commented:
IDBStore is a bit slower, especially subscriptions... looking


**js-store**
```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 7.21 MB/s ±60.3% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 7.68 MB/s ±6.6% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 5.13 MB/s ±2.2% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 3.76 MB/s ±2.1% (5 runs sampled)
read tx 1024x1000 x 8.33 MB/s ±61.6% (6 runs sampled)
read tx 1024x5000 x 8.28 MB/s ±11.4% (5 runs sampled)
scan 1024x1000 x 34.03 MB/s ±243.2% (10 runs sampled)
scan 1024x5000 x 9.91 MB/s ±71.2% (5 runs sampled)
write single byte x 0.00 MB/s ±2147.8% (163 runs sampled)
roundtrip write/subscribe/get x 171.93 ops/sec ±1018.4% (86 runs sampled)
create index 1024x1000 x 10.25 ops/sec ±2.4% (6 runs sampled)
create index 1024x5000 x 1.25 ops/sec ±44.0% (5 runs sampled)
subscription 10 x 6.83 ops/sec ±649.6% (5 runs sampled)
subscription 100 x 5.89 ops/sec ±7.2% (5 runs sampled)
subscription 1000 x 0.88 ops/sec ±62.2% (5 runs sampled)
Done!
```

**main**

```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 6.18 MB/s ±46.8% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 7.91 MB/s ±2.5% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 5.00 MB/s ±2.4% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 3.74 MB/s ±1.6% (5 runs sampled)
read tx 1024x1000 x 8.67 MB/s ±60.2% (6 runs sampled)
read tx 1024x5000 x 8.53 MB/s ±60.0% (5 runs sampled)
scan 1024x1000 x 27.66 MB/s ±185.0% (9 runs sampled)
scan 1024x5000 x 9.93 MB/s ±71.9% (5 runs sampled)
write single byte x 0.00 MB/s ±2073.9% (166 runs sampled)
roundtrip write/subscribe/get x 161.78 ops/sec ±1055.1% (85 runs sampled)
create index 1024x1000 x 10.11 ops/sec ±0.5% (6 runs sampled)
create index 1024x5000 x 2.01 ops/sec ±3.8% (5 runs sampled)
subscription 10 x 15.64 ops/sec ±6.6% (8 runs sampled)
subscription 100 x 5.80 ops/sec ±17.9% (5 runs sampled)
subscription 1000 x 0.88 ops/sec ±56.8% (5 runs sampled)
Done!
```

arv commented:
Perf tests are pretty flaky here... Maybe I need to shut everything down...

arv commented:
I decided to run the perf tests with "everything" shut down:

**js-store**

```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 7.33 MB/s ±52.3% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 8.33 MB/s ±1.1% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 5.22 MB/s ±2.3% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 3.85 MB/s ±1.5% (5 runs sampled)
read tx 1024x1000 x 8.77 MB/s ±59.4% (6 runs sampled)
read tx 1024x5000 x 8.90 MB/s ±62.6% (5 runs sampled)
scan 1024x1000 x 36.17 MB/s ±287.8% (10 runs sampled)
scan 1024x5000 x 10.35 MB/s ±74.9% (5 runs sampled)
write single byte x 0.00 MB/s ±2060.9% (161 runs sampled)
roundtrip write/subscribe/get x 173.65 ops/sec ±1070.8% (87 runs sampled)
create index 1024x1000 x 10.16 ops/sec ±1.7% (6 runs sampled)
create index 1024x5000 x 2.14 ops/sec ±0.7% (5 runs sampled)
subscription 10 x 15.10 ops/sec ±5.9% (8 runs sampled)
subscription 100 x 5.97 ops/sec ±13.6% (5 runs sampled)
subscription 1000 x 0.87 ops/sec ±44.3% (5 runs sampled)
Done!
```

**main**

```
Running benchmarks please wait...
populate 1024x1000 (clean, indexes: 0) x 6.11 MB/s ±34.9% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 8.10 MB/s ±2.0% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 5.06 MB/s ±3.4% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 3.73 MB/s ±1.0% (5 runs sampled)
read tx 1024x1000 x 8.67 MB/s ±60.2% (6 runs sampled)
read tx 1024x5000 x 8.79 MB/s ±60.7% (5 runs sampled)
scan 1024x1000 x 35.38 MB/s ±264.1% (10 runs sampled)
scan 1024x5000 x 10.25 MB/s ±72.1% (5 runs sampled)
write single byte x 0.00 MB/s ±2130.4% (169 runs sampled)
roundtrip write/subscribe/get x 170.27 ops/sec ±1016.7% (89 runs sampled)
create index 1024x1000 x 10.00 ops/sec ±1.4% (6 runs sampled)
create index 1024x5000 x 2.07 ops/sec ±0.7% (5 runs sampled)
subscription 10 x 15.86 ops/sec ±6.3% (8 runs sampled)
subscription 100 x 5.88 ops/sec ±11.6% (5 runs sampled)
subscription 1000 x 0.78 ops/sec ±0.8% (5 runs sampled)
Done!
```

The performance is about the same with a small benefit moving things to JS.

aboodman commented on src/rw-lock.ts:
```diff
@@ -0,0 +1,63 @@
+export class Lock {
+  private _lockP: Promise<unknown> | null = null;
+
+  async lock(): Promise<() => void> {
```
I'm not a fan of exposing this unless it is needed since it is so easy to screw up.

aboodman commented on src/rw-lock.ts:
```diff
@@ -0,0 +1,63 @@
+    return resolve;
+  }
+
+  run<R>(f: () => R | Promise<R>): Promise<R> {
```
Maybe rename `withLock` ?

aboodman commented on src/rw-lock.ts:
```diff
@@ -0,0 +1,63 @@
+import {resolver} from './resolver.js';
+
+export class Lock {
+  private _lockP: Promise<unknown> | null = null;
```
`Promise<void>` really, right?

arv replied:
Doesn't matter. The return value is not used. I'll change it to void since I know it is void

aboodman commented on src/store.ts:
```diff
@@ -0,0 +1,22 @@
+}
+
+export interface Write extends Read {
+  asRead(): Read;
```
I don't expect us to need `asRead` in TS world? Is this only needed for compat with Rust or something?

aboodman commented on src/mem-store.ts:
```diff
@@ -0,0 +1,113 @@
+import {RWLock} from './rw-lock.js';
+import type {Read, Store, Write} from './store.js';
+
+interface Release {
```
DRY - this definition is also in `MemStore` - move up into `Store` ?

aboodman commented on src/mem-store.ts:
```diff
@@ -0,0 +1,113 @@
+      case undefined:
+        return this._map.get(key);
+      default:
+        return v as Uint8Array;
```
I'm surprised that the `as Uint8Array` is necessary. Seems like TS should know it is that type at this point.

aboodman commented on src/idb-store.ts:
```diff
@@ -0,0 +1,244 @@
+      case deleteSentinel:
+        return undefined;
+      case undefined:
+        return (await dbGet(this._tx, key)) as Promise<Uint8Array | undefined>;
```
Cast seems wrong.

arv replied:
Looks right to me

aboodman reviewed (commented):
Nice!

arv committed:
Minor tweaks

arv committed:
More cleanup

arv committed:
more

arv committed:
Move the RWLock into a wrapper class

aboodman commented on src/replicache.ts:
```diff
@@ -291,8 +294,13 @@ export class Replicache<MD extends MutatorDefs = {}>
     // wait for it to finish closing.
     await closingInstances.get(this._name);
 
+    const store = new WrapStore(
```
I don't think we need the in-memory rwlock in the case of IDB. IDB does its own locking internally.

And I don't think we want to expose `WrapStore` -- its only useful in the case of an in-memory store. Other implementations have to do something more clever.

arv replied:
Do we want to rely on the IDB locking mechanism? The current way we end up ensuring things are in an RWLock we control.

WrapStore (renamed to ReleasableStore is not part of the public API)

aboodman replied:
> The current way we end up ensuring things are in an RWLock we control.

I remember talking about this and @phritz arguing for it, but now that I look at it again, I really don't like it - particularly if we are exposing it to users.

A key property of Replicache is correctness and I don't like the squishiness of "the semantics within one tab are a RWLock but across tabs the semantics are ... whatever IDB does".

I think @phritz argument at the time was that the RW Lock was clean and easy to understand, whereas IDB might permit looser models depending on how one reads the spec. But looking at [latest version of the spec](https://www.w3.org/TR/IndexedDB/#transaction-construct), the desired behavior seems very clearly to be a RW lock:

<img width="898" alt="Screen Shot 2021-08-03 at 12 26 36 AM" src="https://user-images.githubusercontent.com/80388/128000723-0f715081-5423-481d-b1c7-9d9c98b4c17c.png">

However, no browsers today implement IDB v3. So the question is whether this change to the spec is meant to be a semantic change or is just a clarification. I am betting it is the latter and that browsers actually implement per-object store RW locks today, since that's the sensible thing.

I think we should confirm this then remove the in-memory RW lock from the IDB path.

arv replied:
I'm also a bit concerned about how to get the current promise based locking behavior. Maybe it is not needed and maybe we do not depend on it but I bet my ass that our tests depends on it.

aboodman commented on src/store.ts:
```diff
@@ -0,0 +1,68 @@
+
+export interface StoreWithRelease {
+  read(): Promise<Read & Release>;
+  write(): Promise<Write & Release>;
```
I think since the semantics of Read.release() and Write.release() are distinct, they should just be part of the Read and Write interfaces directly. Having a shared `Release` interface makes it seem like the `release()` method does the same thing in each case, which it doesn't.

I also think we need to specifically document the rwlock semantics of `Store`.

arv replied:
This is not part of the public API.

And I agree that we need to document the locking behavior. I added that to the JSDoc but it is pretty bare bones.

arv committed:
Move to kv and add tests from Rust

aboodman commented:
I didn’t finish this review yet but wanted to get the comment to you about the direction of the locking stuff.

arv committed:
Feat: Add experimentalKVStore option

And a few refactoring to not pull in test code

arv commented:
I was a bit confused why the tests passed on the bot but I guess it should since the bot ignores the store passed to repc.

arv committed:
cleanup

arv committed:
Store now extends Release

Both MemStore and IDBStore manage their own locking. IDB relieas on
IndexedDB's locking whic hat times can be a bit confusing.

Update the isolation test to be more relaxed and only test certain hard
requirements.

arv committed:
Update JSDoc about locking

arv committed:
Removed dead interface

arv commented on src/kv/store.ts:
```diff
@@ -0,0 +1,50 @@
+ * - Only a single write transaction can be run concurrently.
+ * - A write transaction cannot start until all read transactions have finished.
+ * - If a write transaction is running or waiting, new read transactions have to
+ *   wait until that write transaction is finished.
```
@aboodman Please make sure this makes sense

aboodman commented on src/kv/store.ts:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Key-value store interface that is used with the experimental
```
How about:

```
/**
 * Store defines a transactional key/value store that Replicache stores all data within.
 * 
 * For correct operation of Replicache, implementations of this interface must provide
 * [strict serializable](https://jepsen.io/consistency/models/strict-serializable) transactions.
 *
 * Informally, read and write transactions must behave like a ReadWrite Lock - multiple read
 * transactions are allowed in parallel, or one write. Additionally writes from a transaction
 * must appear all at one, atomically.
 */

aboodman commented on src/kv/store.ts:
```diff
@@ -0,0 +1,50 @@
+  put(key: string, value: Uint8Array): Promise<void>;
+  del(key: string): Promise<void>;
+  commit(): Promise<void>;
+  rollback(): Promise<void>;
```
What is the meaning of `release()` without calling either `commit()` or `rollback()`? Or `commit()` then `rollback()`? 

Perhaps we should just have `commit()` and `release()` .

arv replied:
On the Rust side we use the `Drop` trait to release the lock. This is mostly only needed for Read where there is no commit/rollback.

aboodman commented on src/kv/idb-store.ts:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  release(): void {
+    // Do nothing. We rely on IDB locking.
```
Is there a way to test whether `tx` is open and throw if closed? That would probably be useful for debugging.

arv replied:
I looked into this... There are many code paths in Rust were we neither call commit or rollback. The reason this is not a problem is because rollback is not really needed. IDB auto closes the transaction and memstore just drops the pending map when the Write is dropped.

It could become an issue in the future so maybe I should fix all the Rust code paths?

aboodman replied:
I think actually the other way, if there's no real need for an explicit rollback, let's remove it. That's what I was getting at with my other question about the API design. What if we just have `release()` and `commit()`. I don't think there's a need for an explicit rollback.

Sorry I realize this PR has gone on for awhile. Mostly because I was on break.

aboodman replied:
Looking at Rust now.

arv replied:
> What if we just have `release()` and `commit()`. I don't think there's a need for an explicit rollback.

release is sync and rollback is async.

We cannot make release async.

aboodman commented on src/kv/idb-store.ts:
```diff
@@ -0,0 +1,155 @@
+    this._tx = tx;
+  }
+
+  private _registerTransaction(): void {
```
Couple things confusing to me about how transaction state is tracked:

1. `_registerTransaction` seems misnamed because it doesn't register the transaction with any registry anywhere (I think maybe the Rust code did). All it does is set up some callbacks.
2. There is `_txState` but it's not really a "state", it's a promise that resolves when the transaction closes with its final state. It's really the "end state" of the transaction.
3.  There's also `_transactionState()` which seems related to `_txState` but is different -- it's the current state of the transaction.

What if instead there was `_txState` which wasn't a promise but a normal value of type `WriteState` which reflect the actual state of the tx at this moment. There could also be a `_txStateChange: Promise<void>` that resolves whenever the state changes. At construction time setup the callbacks (currently in `_registerTransaction`) to (a) set the `_txState` correctly and (b) resolve the `_txStateChange` promise and create a new one.

arv replied:
Yeah, this is all legacy. Could use cleanup

arv commented:
Also, make sure you look at the Rust side of this.

arv committed:
Cleanup

arv committed:
Remove rollback WIP

arv committed:
Port more tests from Rust

aboodman commented on src/kv/idb-store.ts:
```diff
@@ -0,0 +1,131 @@
+    await this._onTxEnd;
+
+    if (this._txState !== WriteState.COMMITTED) {
+      throw new Error('Transaction aborted');
```
Can you add to the error message the state the tx was actually in at end?

arv replied:
The error message already includes the word aborted since that is the current state. I'll change the test to make this clearer.

aboodman reviewed (approved):
woo

arv committed:
Change if
